File: lib/src/text/term_matcher.dart
import 'text_normalizer.dart';

/// A dictionary term prepared for matching: a sequence of normalized word
/// tokens plus the id of the owning dictionary entry.
class MatchableTerm {
  MatchableTerm(this.entryId, this.words, {this.partial = false});

  /// Build from a raw term string (e.g. "well-known fact").
  ///
  /// When [partial] is true and the term is a single word, the term will also
  /// match as a sub-word part of longer words (see [TermMatcher]). Partial
  /// matching is silently ignored for multi-word phrases.
  ///
  /// Returns `null` if the term contains no word tokens.
  static MatchableTerm? fromTerm(
    int entryId,
    String term, {
    bool partial = false,
  }) {
    final words = TextNormalizer.tokenize(
      term,
    ).map((t) => TextNormalizer.normalizeToken(t.text)).toList(growable: false);
    if (words.isEmpty) return null;
    return MatchableTerm(entryId, words, partial: partial && words.length == 1);
  }

  final int entryId;
  final List<String> words;

  /// Whether this (single-word) term matches sub-word parts of longer words.
  final bool partial;

  int get wordCount => words.length;
}

/// A located occurrence of a term inside a page's text, as a half-open source
/// range `[start, end)` of code-unit indices into the page's full text.
class TermMatch {
  const TermMatch(this.entryId, this.start, this.end);

  final int entryId;
  final int start;
  final int end;

  int get length => end - start;
}

/// In-memory whole-word / multi-word phrase matcher.
///
/// Matching is done on word boundaries (so "cat" does not match inside
/// "category") which is the right behaviour for a vocabulary highlighter. The
/// matcher tokenizes the page text and, at each token, checks whether any term
/// whose first word equals that token continues to match.
///
/// Construction is O(total term words); a single [findMatches] call is
/// O(page tokens * average candidates per first-word), which in practice is
/// close to linear because most first words have very few candidate terms.
class TermMatcher {
  TermMatcher(Iterable<MatchableTerm> terms) {
    for (final term in terms) {
      if (term.words.isEmpty) continue;
      // Partial single-word terms use the sub-word path below. (That path also
      // covers their whole-word occurrences, so they are not added to the
      // whole-word index as well — which would double-match.)
      if (term.partial && term.wordCount == 1) {
        _partialWords.add(term.words.first);
        _partialEntryIds.add(term.entryId);
        continue;
      }
      (_byFirstWord[term.words.first] ??= <MatchableTerm>[]).add(term);
      if (term.wordCount > _maxWords) _maxWords = term.wordCount;
    }
    // Longer phrases first so that, at a given position, the most specific
    // (longest) term is discovered before shorter ones.
    for (final list in _byFirstWord.values) {
      list.sort((a, b) => b.wordCount.compareTo(a.wordCount));
    }
  }

  final Map<String, List<MatchableTerm>> _byFirstWord = {};
  // Single-word sub-word terms, kept as parallel lists (word, entry id).
  final List<String> _partialWords = [];
  final List<int> _partialEntryIds = [];
  int _maxWords = 0;

  /// Whether the matcher has any terms at all.
  bool get isEmpty => _byFirstWord.isEmpty && _partialWords.isEmpty;

  /// The connectors a normalized token may contain (apostrophe / hyphen). A
  /// sub-word match is accepted when it abuts one of these or a token edge.
  static bool _isConnector(int codeUnit) =>
      codeUnit == 0x2D /* - */ || codeUnit == 0x27 /* ' */;

  /// Find every occurrence of every term in [text].
  ///
  /// Overlapping matches are all returned (e.g. both "bank" and "bank
  /// account"); callers decide how to render or prioritize overlaps.
  List<TermMatch> findMatches(String text) {
    if (isEmpty) return const [];
    final tokens = TextNormalizer.tokenize(text);
    final normalized = List<String>.generate(
      tokens.length,
      (i) => TextNormalizer.normalizeToken(tokens[i].text),
      growable: false,
    );
    final matches = <TermMatch>[];
    for (var i = 0; i < tokens.length; i++) {
      final candidates = _byFirstWord[normalized[i]];
      if (candidates == null) continue;
      for (final term in candidates) {
        final n = term.wordCount;
        if (i + n > tokens.length) continue;
        var ok = true;
        for (var k = 1; k < n; k++) {
          if (normalized[i + k] != term.words[k]) {
            ok = false;
            break;
          }
        }
        if (ok) {
          matches.add(
            TermMatch(term.entryId, tokens[i].start, tokens[i + n - 1].end),
          );
        }
      }
    }
    _findPartialMatches(tokens, normalized, matches);
    return matches;
  }

  /// Sub-word matching for partial single-word terms: a term matches wherever
  /// it occurs inside a token aligned to at least one sub-word boundary — the
  /// token's start/end, or next to a hyphen/apostrophe connector. So
  /// "perturbation" matches the prefix in "perturbations" and the
  /// hyphen-component in "small-perturbation", but not the interior of an
  /// unrelated word like "scatter".
  void _findPartialMatches(
    List<Token> tokens,
    List<String> normalized,
    List<TermMatch> out,
  ) {
    if (_partialWords.isEmpty) return;
    for (var i = 0; i < tokens.length; i++) {
      final tok = normalized[i];
      final tokLen = tok.length;
      final base = tokens[i].start;
      for (var w = 0; w < _partialWords.length; w++) {
        final word = _partialWords[w];
        final wlen = word.length;
        if (wlen == 0 || wlen > tokLen) continue;
        var from = 0;
        while (true) {
          final j = tok.indexOf(word, from);
          if (j < 0) break;
          final endIdx = j + wlen;
          final leftOk = j == 0 || _isConnector(tok.codeUnitAt(j - 1));
          final rightOk =
              endIdx == tokLen || _isConnector(tok.codeUnitAt(endIdx));
          if (leftOk || rightOk) {
            out.add(TermMatch(_partialEntryIds[w], base + j, base + endIdx));
          }
          from = j + 1;
        }
      }
    }
  }

  /// Largest word count among the loaded terms (useful for diagnostics).
  int get maxPhraseWordCount => _maxWords;
}

File: ios-native/Eng/Text/TermMatcher.swift
import Foundation

/// A dictionary term prepared for matching: a sequence of normalized word tokens
/// plus the id of the owning dictionary entry.
struct MatchableTerm {
    let entryId: Int64
    let words: [String]
    /// Whether this (single-word) term also matches sub-word parts of longer words.
    let partial: Bool

    var wordCount: Int { words.count }

    /// Build from a raw term string (e.g. "well-known fact"). `partial` is honored
    /// only for single-word terms. Returns nil if the term has no word tokens.
    static func fromTerm(_ entryId: Int64, _ term: String, partial: Bool = false) -> MatchableTerm? {
        let words = TextNormalizer.tokenize(term).map { TextNormalizer.normalizeToken($0.text) }
        if words.isEmpty { return nil }
        return MatchableTerm(entryId: entryId, words: words, partial: partial && words.count == 1)
    }
}

/// A located occurrence of a term inside a page's text, as a half-open source
/// range `[start, end)` of UTF-16 code-unit indices into the page's full text.
struct TermMatch {
    let entryId: Int64
    let start: Int
    let end: Int
    var length: Int { end - start }
    var range: NSRange { NSRange(location: start, length: end - start) }
}

/// In-memory whole-word / multi-word phrase matcher — a direct port of the
/// Flutter app's `TermMatcher`.
///
/// Matching is done on word boundaries (so "cat" does not match inside
/// "category"). The matcher tokenizes the page text and, at each token, checks
/// whether any term whose first word equals that token continues to match.
final class TermMatcher {
    private var byFirstWord: [String: [MatchableTerm]] = [:]
    // Single-word sub-word terms, kept as parallel arrays (word, entry id).
    private var partialWords: [String] = []
    private var partialEntryIds: [Int64] = []
    private(set) var maxPhraseWordCount = 0

    init<S: Sequence>(_ terms: S) where S.Element == MatchableTerm {
        for term in terms {
            if term.words.isEmpty { continue }
            // Partial single-word terms use the sub-word path below (which also
            // covers their whole-word occurrences), so they are NOT added to the
            // whole-word index as well — that would double-match.
            if term.partial && term.wordCount == 1 {
                partialWords.append(term.words[0])
                partialEntryIds.append(term.entryId)
                continue
            }
            byFirstWord[term.words[0], default: []].append(term)
            if term.wordCount > maxPhraseWordCount { maxPhraseWordCount = term.wordCount }
        }
        // Longer phrases first so the most specific term is discovered first.
        for key in byFirstWord.keys {
            byFirstWord[key]!.sort { $0.wordCount > $1.wordCount }
        }
    }

    var isEmpty: Bool { byFirstWord.isEmpty && partialWords.isEmpty }

    /// A sub-word match is accepted when it abuts a connector (`-` or `'`) or a
    /// token edge.
    private static func isConnector(_ codeUnit: unichar) -> Bool {
        codeUnit == 0x2D /* - */ || codeUnit == 0x27 /* ' */
    }

    /// Find every occurrence of every term in `text`. Overlapping matches are all
    /// returned (e.g. both "bank" and "bank account").
    func findMatches(_ text: String) -> [TermMatch] {
        if isEmpty { return [] }
        let tokens = TextNormalizer.tokenize(text)
        let normalized = tokens.map { TextNormalizer.normalizeToken($0.text) }
        var matches: [TermMatch] = []
        for i in 0..<tokens.count {
            guard let candidates = byFirstWord[normalized[i]] else { continue }
            for term in candidates {
                let n = term.wordCount
                if i + n > tokens.count { continue }
                var ok = true
                var k = 1
                while k < n {
                    if normalized[i + k] != term.words[k] { ok = false; break }
                    k += 1
                }
                if ok {
                    matches.append(TermMatch(entryId: term.entryId,
                                             start: tokens[i].start,
                                             end: tokens[i + n - 1].end))
                }
            }
        }
        findPartialMatches(tokens, normalized, &matches)
        return matches
    }

    /// Sub-word matching for partial single-word terms: a term matches wherever it
    /// occurs inside a token aligned to at least one sub-word boundary — the
    /// token's start/end, or next to a hyphen/apostrophe connector.
    private func findPartialMatches(_ tokens: [Token], _ normalized: [String], _ out: inout [TermMatch]) {
        if partialWords.isEmpty { return }
        // Search on UTF-16 code units so an index into the normalized token maps
        // 1:1 onto the raw page offset (base + j), matching the Dart original.
        let words16 = partialWords.map { Array($0.utf16) }
        for i in 0..<tokens.count {
            let tok = Array(normalized[i].utf16)
            let tokLen = tok.count
            let base = tokens[i].start
            for w in 0..<words16.count {
                let word = words16[w]
                let wlen = word.count
                if wlen == 0 || wlen > tokLen { continue }
                var from = 0
                while true {
                    guard let j = indexOf(word, in: tok, from: from) else { break }
                    let endIdx = j + wlen
                    let leftOk = j == 0 || Self.isConnector(tok[j - 1])
                    let rightOk = endIdx == tokLen || Self.isConnector(tok[endIdx])
                    if leftOk || rightOk {
                        out.append(TermMatch(entryId: partialEntryIds[w],
                                             start: base + j, end: base + endIdx))
                    }
                    from = j + 1
                }
            }
        }
    }

    /// First index >= `from` at which `needle` occurs in `haystack` (UTF-16 units).
    private func indexOf(_ needle: [unichar], in haystack: [unichar], from: Int) -> Int? {
        let n = needle.count, h = haystack.count
        if n == 0 || n > h { return nil }
        var i = from
        while i + n <= h {
            var k = 0
            while k < n && haystack[i + k] == needle[k] { k += 1 }
            if k == n { return i }
            i += 1
        }
        return nil
    }
}
